zebra-network/src/peer/handshake.rs

```diff
@@ -1017,6 +959,78 @@ where
     }
 }
 
+/// Send periodical heartbeats to `server_tx`, and update the peer status through
+/// `heartbeat_ts_collector`.
+///
+/// # Correctness
+///
+/// To prevent hangs:
+/// - every await that depends on the network must have a timeout (or interval)
+/// - every error/shutdown must update the address book state and return
+///
+/// The address book state can be updated via `ClientRequest.tx`, or the
+/// heartbeat_ts_collector.
+///
+/// Returning from the spawned closure terminates the connection's heartbeat task.
+async fn send_periodic_heartbeats(
+    connected_addr: ConnectedAddr,
+    remote_services: PeerServices,
+    mut shutdown_rx: oneshot::Receiver<CancelHeartbeatTask>,
+    mut server_tx: mpsc::Sender<ClientRequest>,
+    mut heartbeat_ts_collector: mpsc::Sender<MetaAddrChange>,
+) {
+    use futures::future::Either;
+
+    let mut interval_stream =
+        IntervalStream::new(tokio::time::interval(constants::HEARTBEAT_INTERVAL));
+
+    loop {
+        let shutdown_rx_ref = Pin::new(&mut shutdown_rx);
+
+        // CORRECTNESS
+        //
+        // Currently, select prefers the first future if multiple
+        // futures are ready.
+        //
+        // Starvation is impossible here, because interval has a
+        // slow rate, and shutdown is a oneshot. If both futures
+        // are ready, we want the shutdown to take priority over
+        // sending a useless heartbeat.
+        if matches!(
+            future::select(shutdown_rx_ref, interval_stream.next()).await,
+            Either::Left(_)
+        ) {
+            tracing::trace!("shutting down due to Client shut down");
+            if let Some(book_addr) = connected_addr.get_address_book_addr() {
+                // awaiting a local task won't hang
+                let _ = heartbeat_ts_collector
+                    .send(MetaAddr::new_shutdown(&book_addr, remote_services))
+                    .await;
+            }
+            return;
+        }
+
+        // We've reached another heartbeat interval without
+        // shutting down, so do a heartbeat request.
+        //
+        // TODO: await heartbeat and shutdown. The select
+        // function needs pinned types, but pinned generics
+        // are hard (#1678)
```

Comment:
Related ticket update: let's try to do this fix before the network audit (but not in this PR)

```suggestion
        // TODO: await heartbeat and shutdown (#3254)
```